go/exosocket/index.go

```diff
@@ -0,0 +1,106 @@
+package exosocket
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Originate/exocom/go/structs"
+	"github.com/Originate/exocom/go/utils"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+)
+
+// ExoSocket is the low level API to Exocom
+type ExoSocket struct {
+	config                      Config
+	shouldReconnectOnDisconnect bool
+	socket                      *websocket.Conn
+	messageChannel              chan structs.Message
+}
+
+// NewExoSocket returns an ExoSocket with the given config
+func NewExoSocket(config Config) *ExoSocket {
+	return &ExoSocket{config: config}
+}
+
+// Connect brings an ExoSocket instance online
+func (e *ExoSocket) Connect() error {
+	if e.config.Port == 0 {
+		e.config.Port = 80
+	}
+	exocomURL := fmt.Sprintf("ws://%s:%d/services", e.config.Host, e.config.Port)
+	socket, err := utils.ConnectWithRetry(exocomURL, 100)
+	if err != nil {
+		return err
+	}
+	e.shouldReconnectOnDisconnect = true
+	e.socket = socket
+	e.messageChannel = make(chan structs.Message)
+	go e.listenForMessages()
+	_, err = e.Send(MessageOptions{
+		Name: "exocom.register-service",
+	})
+	return err
+}
+
+// Close takes this ExoRelay instance offline
+func (e *ExoSocket) Close() error {
+	e.shouldReconnectOnDisconnect = false
+	return e.socket.Close()
+}
+
+// GetMessageChannel returns a channel which can be used read incoming messages
+func (e *ExoSocket) GetMessageChannel() chan structs.Message {
+	return e.messageChannel
+}
+
+// Send sends a message with the given options
+func (e *ExoSocket) Send(options MessageOptions) (*structs.Message, error) {
+	var activityID string
+	if e.socket == nil {
+		return &structs.Message{}, errors.New("ExoRelay#Send not connected to Exocom")
+	}
+	id := uuid.NewV4().String()
+	if options.Name == "" {
+		return &structs.Message{}, errors.New("ExoRelay#Send cannot send empty messages")
+	}
+	if options.ActivityID == "" {
+		activityID = uuid.NewV4().String()
+	} else {
+		activityID = options.ActivityID
+	}
+	message := &structs.Message{
+		ID:         id,
+		Name:       options.Name,
+		Payload:    options.Payload,
+		ActivityID: activityID,
+		Sender:     e.config.Role,
+		Auth:       options.Auth,
+		IsSecurity: options.IsSecurity,
+	}
+	serializedBytes, err := json.Marshal(message)
+	if err != nil {
+		return &structs.Message{}, err
+	}
+	return message, e.socket.WriteMessage(websocket.TextMessage, serializedBytes)
+}
+
+// Helpers
+
+func (e *ExoSocket) listenForMessages() {
+	utils.ListenForMessages(e.socket, func(message structs.Message) error {
+		e.messageChannel <- message
```

Comment:
I'm not entirely sure this works. The underlying channel can be replaced by `Connect` when reconnecting. It's unclear if the value captured by this closure would get updated as well

Reply:
Great catch. Moved the channel creation into the constructor so it is no longer re-created on each connect